### PR TITLE
enhancement: Update StructuredGraph rebuilding logic

### PR DESCRIPTION
**Pull request by rish691**
https://github.com/hypertrace/hypertrace/issues/244

### CODE REVIEW

File: hypertrace-trace-enricher/hypertrace-trace-enricher-api/src/main/java/org/hypertrace/traceenricher/trace/util/ApiTraceGraphBuilder.java [hunk 1]
```diff
@@ -2,25 +2,23 @@
...
   private static final Logger LOG = LoggerFactory.getLogger(ApiTraceGraphBuilder.class);
 
-  private static ThreadLocal<ApiTraceGraph> cachedGraph = new ThreadLocal<>();
-  private static ThreadLocal<StructuredTrace> cachedTrace = new ThreadLocal<>();
+  private static final ThreadLocal<ApiTraceGraph> cachedGraph = new ThreadLocal<>();
+  private static final ThreadLocal<StructuredTrace> cachedTrace = new ThreadLocal<>();
 
   public static ApiTraceGraph buildGraph(StructuredTrace trace) {
-    if (!GraphBuilderUtil.isSameStructuredTrace(cachedTrace.get(), trace)) {
+    if (GraphBuilderUtil.isStructuredTraceChanged(cachedTrace.get(), trace)) {
```
Comment (kotharironak) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-api/src/main/java/org/hypertrace/traceenricher/trace/util/ApiTraceGraphBuilder.java:16]: I thin, ApiTraceGraph is independent of entitygraph. So, do we have to built it again?

File: hypertrace-trace-enricher/hypertrace-trace-enricher-api/src/main/java/org/hypertrace/traceenricher/trace/util/GraphBuilderUtil.java [hunk 2]
```diff
@@ -7,34 +7,55 @@
...
+        || isTraceEntitiesChanged(cachedTrace, trace);
+  }
 
-    // trace internally changed (full trace comparision is costly, so we are doing only with
+  /** Check if the events or theirs edges has changed */
+  static boolean isTraceEventsChanged(StructuredTrace cachedTrace, StructuredTrace trace) {
+    // trace events internally changed (full trace comparison is costly, so we are doing only with
     // required fields)
-    if (cachedTrace.getEntityList().size() != trace.getEntityList().size()
+    if (null == cachedTrace
```
Comment (kotharironak) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-api/src/main/java/org/hypertrace/traceenricher/trace/util/GraphBuilderUtil.java:35]: Shall we replace this with isDifferentTrace only?

File: hypertrace-trace-enricher/hypertrace-trace-enricher-api/src/main/java/org/hypertrace/traceenricher/trace/util/ApiTraceGraphBuilder.java (hunk 1 above)
Comment (rish691) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-api/src/main/java/org/hypertrace/traceenricher/trace/util/ApiTraceGraphBuilder.java:16]: Yeah, I observed that. But thought of taking up that refactor separately.

File: hypertrace-trace-enricher/hypertrace-trace-enricher-api/src/main/java/org/hypertrace/traceenricher/trace/util/GraphBuilderUtil.java (hunk 2 above)
Comment (rish691) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-api/src/main/java/org/hypertrace/traceenricher/trace/util/GraphBuilderUtil.java]: Done

**Review by kotharironak — APPROVED** [review]
lgtm. Minor comments

**Pull request merged**